File: samples/python/object_detection_yolo.py
import json
import logging as log
import os
import sys
import time
from datetime import datetime, timezone
from pathlib import Path

import cv2
import numpy as np
import openvino as ov
import yaml

# python samples/python/object_detection_yolo.py models/yolo/yolo26n_openvino_model/yolo26n.xml data/images/person/person_detection.png GPU
# python samples/python/object_detection_yolo.py models/yolo/yolo26n_openvino_model/yolo26n.xml data/images/person/person-bicycle-car-detection.bmp GPU
def main():
    log.basicConfig(format='[ %(levelname)s ] %(message)s', level=log.INFO, stream=sys.stdout)

    # Parsing and validation of input arguments
    if len(sys.argv) != 4:
        log.info(f'Usage: {sys.argv[0]} <path_to_model> <path_to_image> <device_name>')
        return 1

    model_path = sys.argv[1]
    image_path = sys.argv[2]
    device_name = sys.argv[3]

    CONF_THRESHOLD = 0.5
    INPUT_SIZE = 640  # YOLO model fixed input size

    # Load class names from metadata.yaml alongside the model
    metadata_path = Path(model_path).parent / 'metadata.yaml'
    with open(metadata_path) as f:
        metadata = yaml.safe_load(f)
    class_names = metadata.get('names', {})

# --------------------------- Step 1. Initialize OpenVINO Runtime Core ------------------------------------------------
    log.info('Creating OpenVINO Runtime Core')
    core = ov.Core()

# --------------------------- Step 2. Read a model --------------------------------------------------------------------
    log.info(f'Reading the model: {model_path}')
    model = core.read_model(model_path)

    if len(model.inputs) != 1:
        log.error('Sample supports only single input topologies')
        return -1

# --------------------------- Step 3. Set up input --------------------------------------------------------------------
    # Read input image and keep original for drawing
    image = cv2.imread(image_path)
    orig_h, orig_w = image.shape[:2]

    # Resize to model's fixed input size (640x640) — do NOT reshape the model,
    # because transformer attention layers have incompatible fixed internal shapes.
    resized = cv2.resize(image, (INPUT_SIZE, INPUT_SIZE))

    # Convert BGR -> RGB, HWC -> NCHW, normalize to [0, 1] as float32
    input_tensor = cv2.cvtColor(resized, cv2.COLOR_BGR2RGB).astype(np.float32) / 255.0
    input_tensor = np.transpose(input_tensor, (2, 0, 1))   # HWC -> CHW
    input_tensor = np.expand_dims(input_tensor, axis=0)    # CHW -> NCHW

# ---------------------------Step 4. Loading model to the device-------------------------------------------------------
    log.info('Loading the model to the plugin')
    compiled_model = core.compile_model(model, device_name)

# --------------------------- Step 5. Create infer request and do inference synchronously -----------------------------
    log.info('Starting inference in synchronous mode')
    t0 = time.perf_counter()
    results = compiled_model.infer_new_request({0: input_tensor})
    inference_ms = round((time.perf_counter() - t0) * 1000, 2)

# ---------------------------Step 6. Process output--------------------------------------------------------------------
    # YOLO exported model output: [1, 300, 6] — NMS already applied by the model
    # Each row: [x1, y1, x2, y2, score, class_id] in INPUT_SIZE pixel space
    predictions = next(iter(results.values()))  # [1, 300, 6]
    detections = np.squeeze(predictions, axis=0)  # [300, 6]

    # Scale factors from model input size back to original image size
    scale_x = orig_w / INPUT_SIZE
    scale_y = orig_h / INPUT_SIZE

    # Although NMS is built into the model, the threshold my be set too loosely, resulting in highly overlapping bounding boxed being retained simultaneously.
    # The model utilizes float16 or lower-precision arithmetic, which can lead to two nearly identical anchor boxes both passing the filtering stage. This is a common issue with YOLO models, especially when the confidence threshold is set to a value that allows for more detections.
    # IoU filtering, when the ratio of the overlapping area between two bounding boxes to their union area exceeds 0.45, the box with the higher confidence score is retained, while the one with the lower score is discarded. This helps to reduce the number of duplicate detections and improve the overall quality of the results.
    NMS_IOU_THRESHOLD = 0.45

    # First pass: collect all candidates above confidence threshold
    candidates = []
    for det in detections:
        x1, y1, x2, y2, score, class_id = det
        if score < CONF_THRESHOLD:
            continue
        class_id = int(class_id)
        xmin = int(x1 * scale_x)
        ymin = int(y1 * scale_y)
        xmax = int(x2 * scale_x)
        ymax = int(y2 * scale_y)
        candidates.append({
            'class_id': class_id,
            'label': class_names.get(class_id, f'class_{class_id}'),
            'confidence': float(score),
            # cv2.dnn.NMSBoxes expects [x, y, w, h]
            'box_xywh': [xmin, ymin, xmax - xmin, ymax - ymin],
            'bbox': {'xmin': xmin, 'ymin': ymin, 'xmax': xmax, 'ymax': ymax},
        })

    # Second pass: apply NMS per class to suppress overlapping duplicates
    detection_results = []
    if candidates:
        boxes = [c['box_xywh'] for c in candidates]
        scores = [c['confidence'] for c in candidates]
        keep_indices = cv2.dnn.NMSBoxes(boxes, scores, CONF_THRESHOLD, NMS_IOU_THRESHOLD)
        for idx in keep_indices:
            c = candidates[idx]
            detection_results.append({
                'id': len(detection_results),
                'class_id': c['class_id'],
                'label': c['label'],
                'confidence': round(c['confidence'], 4),
                'bbox': c['bbox'],
            })
            b = c['bbox']
            cv2.rectangle(image, (b['xmin'], b['ymin']), (b['xmax'], b['ymax']), (0, 255, 0), 2)

    output = {
        'timestamp': datetime.now(timezone.utc).isoformat(),
        'image': image_path,
        'model': model_path,
        'device': device_name,
        'inference_ms': inference_ms,
        'num_detections': len(detection_results),
        'detections': detection_results,
    }
    log.info('Detection results:\n' + json.dumps(output, indent=2))

    cv2.imwrite('out.bmp', image)

    if os.path.exists('out.bmp'):
        log.info('Image out.bmp was created!')
    else:
        log.error('Image out.bmp was not created. Check your permissions.')

# ----------------------------------------------------------------------------------------------------------------------
    log.info('This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool\n')
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: samples/cpp/object_detection_yolo.cpp
#include <chrono>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

#include <opencv2/dnn.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include "openvino/openvino.hpp"

// ---------------------------------------------------------------------------
// Parse the 'names' section from a YOLO metadata.yaml.
// Expected format:
//   names:
//     0: person
//     1: bicycle
// ---------------------------------------------------------------------------
static std::map<int, std::string> load_class_names(const std::string& yaml_path) {
    std::map<int, std::string> names;
    std::ifstream f(yaml_path);
    if (!f.is_open())
        return names;

    bool in_names = false;
    std::string line;
    while (std::getline(f, line)) {
        if (line.find("names:") != std::string::npos) {
            in_names = true;
            continue;
        }
        if (!in_names)
            continue;

        // Stop at the next top-level key (no leading whitespace, not a digit line)
        if (!line.empty() && line[0] != ' ' && line[0] != '\t')
            break;

        size_t colon = line.find(':');
        if (colon == std::string::npos)
            continue;

        std::string key_str = line.substr(0, colon);
        key_str.erase(0, key_str.find_first_not_of(" \t"));
        if (key_str.empty() || !std::isdigit(static_cast<unsigned char>(key_str[0])))
            continue;

        int id = std::stoi(key_str);
        std::string val = line.substr(colon + 1);
        val.erase(0, val.find_first_not_of(" \t"));
        if (!val.empty() && val.back() == '\r')
            val.pop_back();
        names[id] = val;
    }
    return names;
}

// ---------------------------------------------------------------------------
// Minimal JSON helpers — no third-party library required.
// ---------------------------------------------------------------------------
static std::string json_str(const std::string& s) {
    return "\"" + s + "\"";
}

static std::string utc_timestamp() {
    auto now = std::chrono::system_clock::now();
    std::time_t t = std::chrono::system_clock::to_time_t(now);
    std::tm tm_utc{};
    gmtime_r(&t, &tm_utc);
    char buf[32];
    std::strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S+00:00", &tm_utc);
    return buf;
}

// ---------------------------------------------------------------------------
int main(int argc, char* argv[]) {
    try {
        if (argc != 4) {
            std::cout << "Usage: " << argv[0]
                      << " <path_to_model> <path_to_image> <device_name>" << std::endl;
            std::cout << "Example: " << argv[0]
                      << " models/yolo/yolo26n_openvino_model/yolo26n.xml data/images/person/person_detection.png GPU"
                      << std::endl;
            return EXIT_FAILURE;
        }

        const std::string model_path{argv[1]};
        const std::string image_path{argv[2]};
        const std::string device_name{argv[3]};

        constexpr float CONF_THRESHOLD    = 0.5f;
        constexpr float NMS_IOU_THRESHOLD = 0.45f;
        constexpr int   INPUT_SIZE        = 640;

        // Load class names from metadata.yaml alongside the model
        const std::string dir =
            model_path.substr(0, model_path.find_last_of("/\\") + 1);
        const std::map<int, std::string> class_names =
            load_class_names(dir + "metadata.yaml");

// --------------------------- Step 1. Initialize OpenVINO Runtime Core ---------------
        std::cout << "[ INFO ] Creating OpenVINO Runtime Core" << std::endl;
        ov::Core core;

// --------------------------- Step 2. Read a model -----------------------------------
        std::cout << "[ INFO ] Reading the model: " << model_path << std::endl;
        std::shared_ptr<ov::Model> model = core.read_model(model_path);

        if (model->inputs().size() != 1) {
            std::cerr << "[ ERROR ] Sample supports only single input topologies"
                      << std::endl;
            return EXIT_FAILURE;
        }

// --------------------------- Step 3. Set up input -----------------------------------
        cv::Mat image = cv::imread(image_path);
        if (image.empty()) {
            std::cerr << "[ ERROR ] Cannot read image: " << image_path << std::endl;
            return EXIT_FAILURE;
        }
        const int orig_h = image.rows;
        const int orig_w = image.cols;

        // Resize to INPUT_SIZE x INPUT_SIZE — do NOT reshape the model,
        // because transformer attention layers have incompatible fixed internal shapes.
        cv::Mat resized;
        cv::resize(image, resized, cv::Size(INPUT_SIZE, INPUT_SIZE));

        // Convert BGR -> RGB, normalize to [0, 1] as float32
        cv::Mat rgb;
        cv::cvtColor(resized, rgb, cv::COLOR_BGR2RGB);
        rgb.convertTo(rgb, CV_32F, 1.0 / 255.0);

        // HWC -> NCHW: split per-channel and lay out contiguously
        std::vector<cv::Mat> ch(3);
        cv::split(rgb, ch);
        std::vector<float> input_data(1 * 3 * INPUT_SIZE * INPUT_SIZE);
        for (int c = 0; c < 3; ++c) {
            const float* src = reinterpret_cast<const float*>(ch[c].data);
            std::copy(src, src + INPUT_SIZE * INPUT_SIZE,
                      input_data.data() + c * INPUT_SIZE * INPUT_SIZE);
        }

// --------------------------- Step 4. Load model to device ---------------------------
        std::cout << "[ INFO ] Loading the model to the plugin" << std::endl;
        ov::CompiledModel compiled_model = core.compile_model(model, device_name);

// --------------------------- Step 5. Inference synchronously ------------------------
        std::cout << "[ INFO ] Starting inference in synchronous mode" << std::endl;
        ov::InferRequest infer_request = compiled_model.create_infer_request();

        ov::Tensor input_tensor(
            ov::element::f32,
            {1, 3, static_cast<size_t>(INPUT_SIZE), static_cast<size_t>(INPUT_SIZE)},
            input_data.data());
        infer_request.set_input_tensor(input_tensor);

        auto t0 = std::chrono::high_resolution_clock::now();
        infer_request.infer();
        auto t1 = std::chrono::high_resolution_clock::now();
        const double inference_ms =
            std::chrono::duration<double, std::milli>(t1 - t0).count();

// --------------------------- Step 6. Process output ---------------------------------
        // YOLO exported model output: [1, 300, 6] — NMS already applied by the model
        // Each row: [x1, y1, x2, y2, score, class_id] in INPUT_SIZE pixel space
        const ov::Tensor output_tensor = infer_request.get_output_tensor();
        const ov::Shape  out_shape     = output_tensor.get_shape();  // [1, 300, 6]
        const size_t     num_boxes     = out_shape[1];
        const size_t     box_size      = out_shape[2];  // 6
        const float*     raw           = output_tensor.data<const float>();

        const float scale_x = static_cast<float>(orig_w) / INPUT_SIZE;
        const float scale_y = static_cast<float>(orig_h) / INPUT_SIZE;

        // Although NMS is built into the model, the threshold may be set too loosely,
        // resulting in highly overlapping boxes being retained simultaneously (common
        // with float16 anchor arithmetic).  Apply a second NMS pass here.
        struct Candidate {
            int         class_id;
            std::string label;
            float       confidence;
            cv::Rect    bbox;   // in original image coordinates
        };

        // First pass: collect all candidates above confidence threshold
        std::vector<Candidate>  candidates;
        std::vector<cv::Rect>   boxes_vec;
        std::vector<float>      scores_vec;

        for (size_t i = 0; i < num_boxes; ++i) {
            const float* det = raw + i * box_size;
            const float score    = det[4];
            if (score < CONF_THRESHOLD)
                continue;

            const int xmin = static_cast<int>(det[0] * scale_x);
            const int ymin = static_cast<int>(det[1] * scale_y);
            const int xmax = static_cast<int>(det[2] * scale_x);
            const int ymax = static_cast<int>(det[3] * scale_y);
            const int cid  = static_cast<int>(det[5]);

            auto it = class_names.find(cid);
            const std::string label =
                (it != class_names.end()) ? it->second : "class_" + std::to_string(cid);

            cv::Rect bbox(xmin, ymin, xmax - xmin, ymax - ymin);
            candidates.push_back({cid, label, score, bbox});
            boxes_vec.push_back(bbox);
            scores_vec.push_back(score);
        }

        // Second pass: apply NMS to suppress overlapping duplicates
        std::vector<int> keep_indices;
        if (!boxes_vec.empty())
            cv::dnn::NMSBoxes(boxes_vec, scores_vec,
                              CONF_THRESHOLD, NMS_IOU_THRESHOLD, keep_indices);

        // Draw bounding boxes on the original image and build JSON output
        std::ostringstream detections_json;
        for (size_t i = 0; i < keep_indices.size(); ++i) {
            const Candidate& c = candidates[keep_indices[i]];
            cv::rectangle(image, c.bbox, cv::Scalar(0, 255, 0), 2);

            if (i > 0)
                detections_json << ",\n";
            detections_json
                << "      {\n"
                << "        \"id\": "         << i                          << ",\n"
                << "        \"class_id\": "   << c.class_id                 << ",\n"
                << "        \"label\": "      << json_str(c.label)          << ",\n"
                << "        \"confidence\": " << std::fixed << std::setprecision(4)
                                              << c.confidence               << ",\n"
                << "        \"bbox\": {\n"
                << "          \"xmin\": "     << c.bbox.x                   << ",\n"
                << "          \"ymin\": "     << c.bbox.y                   << ",\n"
                << "          \"xmax\": "     << c.bbox.x + c.bbox.width    << ",\n"
                << "          \"ymax\": "     << c.bbox.y + c.bbox.height   << "\n"
                << "        }\n"
                << "      }";
        }

        std::ostringstream output_json;
        output_json
            << "{\n"
            << "  \"timestamp\": "      << json_str(utc_timestamp())        << ",\n"
            << "  \"image\": "          << json_str(image_path)             << ",\n"
            << "  \"model\": "          << json_str(model_path)             << ",\n"
            << "  \"device\": "         << json_str(device_name)            << ",\n"
            << "  \"inference_ms\": "   << std::fixed << std::setprecision(2)
                                        << inference_ms                     << ",\n"
            << "  \"num_detections\": " << keep_indices.size()              << ",\n"
            << "  \"detections\": [\n"  << detections_json.str()            << "\n"
            << "  ]\n"
            << "}";

        std::cout << "[ INFO ] Detection results:\n" << output_json.str() << std::endl;

        cv::imwrite("out.bmp", image);
        if (std::ifstream("out.bmp").good())
            std::cout << "[ INFO ] Image out.bmp was created!" << std::endl;
        else
            std::cerr << "[ ERROR ] Image out.bmp was not created. Check your permissions."
                      << std::endl;

    } catch (const std::exception& ex) {
        std::cerr << ex.what() << std::endl;
        return EXIT_FAILURE;
    }

    std::cout << "\nThis sample is an API example, for any performance measurements "
                 "please use the dedicated benchmark_app tool"
              << std::endl;
    return EXIT_SUCCESS;
}
